homeassistant/components/device_tracker/owntracks_http.py

```diff
@@ -4,52 +4,122 @@
 For more details about this platform, please refer to the documentation at
 https://home-assistant.io/components/device_tracker.owntracks_http/
 """
+import json
+import logging
 import re
 
-from aiohttp.web_exceptions import HTTPInternalServerError
+from aiohttp.web import Response
+import voluptuous as vol
 
-from homeassistant.components.http import HomeAssistantView
+from homeassistant.components.device_tracker.owntracks import (  # NOQA
+    PLATFORM_SCHEMA, REQUIREMENTS, async_handle_message, context_from_config)
+from homeassistant.const import CONF_WEBHOOK_ID
+from homeassistant.core import callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect, async_dispatcher_send)
 
-# pylint: disable=unused-import
-from .owntracks import (  # NOQA
-    REQUIREMENTS, PLATFORM_SCHEMA, context_from_config, async_handle_message)
+DEPENDENCIES = ['webhook']
 
+_LOGGER = logging.getLogger(__name__)
 
-DEPENDENCIES = ['http']
+EVENT_RECEIVED = 'owntracks_http_webhook_received'
+EVENT_RESPONSE = 'owntracks_http_webhook_response_'
+
+DOMAIN = 'device_tracker.owntracks_http'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_WEBHOOK_ID): cv.string
+})
+
+
+async def async_setup(hass, config):
+    """Set up the OwnTracks HTTP service component."""
+    return True
 
 
 async def async_setup_scanner(hass, config, async_see, discovery_info=None):
-    """Set up an OwnTracks tracker."""
+    """Set up OwnTracks HTTP component."""
+    hass.components.webhook.async_register(
+        config['webhook_id'], handle_webhook)
+
     context = context_from_config(async_see, config)
 
-    hass.http.register_view(OwnTracksView(context))
+    subscription = context.mqtt_topic
+    topic = re.sub('/#$', '', subscription)
+
+    async def async_handle_event(data):
+        """Handle incoming webhook post from OwnTracks."""
+        message = data['message']
+        message['topic'] = '{}/{}/{}'.format(topic, data['user'],
+                                             data['device'])
+
+        response = dict()
+
+        try:
+            await async_handle_message(hass, context, message)
+            response = {'body': json.dumps([]), 'status': 200}
+        except ValueError:
+            response = {'body': json.dumps({'error': 'Error parsing JSON'}),
+                        'status': 500}
+
+        async_dispatcher_send(hass, EVENT_RESPONSE + str(message['tst']),
```

Comment:
I don't see any reason why this needs to use a dispatcher. It should all be done inside the webhook handler.

Reply:
I can't find a way to access `async_see` and `config` from the webhook handler to initialize the context in https://github.com/home-assistant/home-assistant/pull/17034/files/ccec3d6a60ac14f4af9e9854f1a378cfed103c1b#diff-8cd102f735f57c74c7f9ebe438dce2e7R46 

Without the owntracks context I can't get the MQTT topic. Maybe I'm missing something and I can extract that data from the `hass` object directly, as I dislike this back-and-forth between the callbacks too.